### PR TITLE
Enable selection set analysis before field resolution starts

### LINKED ISSUE #1695: QueryTraverser only descends into FragmentSpreads once
When the same fragment is used multiple times, `QueryTraverser` descends into its field only once.

A simple example:

```graphql
{
  users {
    ...userInfo
    friends {
      ...userInfo
    }
  }

  fragment userInfo on User {
    name
  }
}
```

On a query such as this, `QueryTraverser` will visit the `name` field only once (and will visit the `...userInfo` spread twice, as expected). This is distinctly different from the behavior one would get for a semantically equivalent query with no fragments:

```graphql
{
  users {
    name
    friends {
      name
    }
  }
}
```
One immediate implication is on query complexity analysis. A (malicious?) query with recursive fragment spreads will get a score underestimated by orders of magnitude.

As an example, `MaxQueryComplexityInstrumentation` will calculate the complexity of the full introspection query to be 45, while in reality over a 100 fields are requested. And that query isn't intentionally crafted to trick the calculation with large, recursive, fragment spreads.

### Investigation

The reason why this happens is that `QueryTraverser` [in this line](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/analysis/QueryTraverser.java#L158) considers `FragmentDefinition` the only child of each `FragmentSpread`. Since both spreads have the same definitions, when the traverser encounters the 2nd spread, it will falsely believe ([in this line](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/util/DefaultTraverserContext.java#L149)) its children (the definition) have already been visited and will not descend.

Happy to help fixing this, but I'd need guidance on the approach...

### PR DESCRIPTION
**Pull request by kaqqao**
Currently, complexity analysis as implemented by graphql-java (using `QueryTraverser`) is inaccurate and can severely underestimate the complexity of recursive queries with fragment spreads (as described in the long-standing issue #1695). Additionally (and IMHO more importantly), it is impossible to build more accurate analysis using the public APIs.

While `ExecutableNormalizedOperation` enables accurate (and customizable) analysis of the entire (possibly multi-root) query before any field resolution starts, it is an internal class. On the other hand, `DataFetchingFieldSelectionSet` exposes all the necessary info (in an even friendlier way!), and is public, but is only created for individual fields once field resolution starts, and thus it does not allow early analysis (e.g. from `Instrumentation#beginExecuteOperation`) of the entire query (i.e. of all the roots) in order to completely prevent execution if a rule is violated.

With this PR, a list of `DataFetchingFieldSelectionSet`s is (lazily) created for all root fields before field resolution starts. This list is then made accessible via `ExecutionContext`, so it is possible to retrieve it from an instrumentation at any stage where `ExecutionContext` exists, such as `beginExecuteOperation`. This way all the interesting information from the normalized query tree (i.e. `ExecutableNormalizedOperation`) is exposed to complexity analysis as early as the tree itself exists, but without exposing either the tree or any other private API.

I am not exactly sure this is the correct approach, but might at least serve to start the conversation.

Fixes #1695

### CODE REVIEW

File: src/main/java/graphql/execution/ExecutionContext.java [hunk 1]
```diff
@@ -80,7 +84,17 @@ public class ExecutionContext {
         this.errors.set(builder.errors);
         this.localContext = builder.localContext;
         this.executionInput = builder.executionInput;
-        queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.queryTree = FpKit.interThreadMemoize(() -> ExecutableNormalizedOperationFactory.createExecutableNormalizedOperation(graphQLSchema, operationDefinition, fragmentsByName, coercedVariables));
+        this.rootDataFetchingFieldSelectionSet = FpKit.intraThreadMemoize(() -> collectRootDataFetchingFieldSelectionSets(queryTree, builder.graphQLSchema));
```
Comment (kaqqao) [inline src/main/java/graphql/execution/ExecutionContext.java:88]: Intra-thread as the creating `DataFetchingFieldSelectionSet` isn't an expensive operation in itself to warrant synchronized access. But that might be a wrong conclusion.

File: src/main/java/graphql/execution/ExecutionContext.java [hunk 2]
```diff
@@ -80,7 +84,17 @@ public class ExecutionContext {
...
+    }
+
+    @NotNull
+    private List<DataFetchingFieldSelectionSet> collectRootDataFetchingFieldSelectionSets(
+            Supplier<ExecutableNormalizedOperation> queryTree, GraphQLSchema schema) {
+        return ImmutableKit.map(queryTree.get().getTopLevelFields(), root -> DataFetchingFieldSelectionSetImpl.newCollector(
+                schema,
+                root.getFieldDefinitions(schema).get(0).getType(), //root fields always have a single parent
```
Comment (kaqqao) [inline src/main/java/graphql/execution/ExecutionContext.java:96]: This is, I believe, safe. All root fields have a singular parent, introspection fields included. Right?

**Comment by bbakerman** [comment]
I am supportive of this - eg making the top level field selection sets available

However this PR needs way more tests to be acceptable

The magic fields (`__typename`, `__schema` etc..) need to be included to make sure this works

And we need some complex queries with nest levels to just make sure we build what we think we have built.

**Review by bbakerman — CHANGES_REQUESTED** [review]
See my comments on what is needed for this to be acceptable

**Comment by kaqqao** [comment]
My main intention with this PR was to expose the info from ENF, without exposing ENF itself. With ENF becoming public (🥳), I don't think adding this is particularly useful, so I'll close the PR. But if you disagree, I'll clean it up as you asked and reopen.

**Pull request closed without merging**